### PR TITLE
Forget roots and favs

### PR DESCRIPTION
**Pull request by shubertm**

### CODE REVIEW

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt [hunk 1]
```diff
@@ -119,11 +113,57 @@ class FoldersRepo(private val appCtx: Context) {
...
-        favsFile.writeText(klaxon.toJsonString(jsonFavs))
+        writeFavorites(root)
+    }
+
+    suspend fun forgetRoot(root: Path) =
+        withContext(Dispatchers.IO) {
+            if (folders.containsKey(root)) {
+                val mFolders = folders.toMutableMap()
+                mFolders.remove(root)
+                folders = mFolders
```
Comment (kirillt) [inline arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Probably we can just perform an immutable deletion.

Comment (kirillt) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/minus.html

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt [hunk 2]
```diff
@@ -119,11 +113,57 @@ class FoldersRepo(private val appCtx: Context) {
...
+            }
+        }
+
+    suspend fun forgetFavorite(root: Path, favorite: Path) =
+        withContext(Dispatchers.IO) {
+            if (folders.containsKey(root)) {
+                val favorites = folders[root]
+                val favRelative = root.relativize(favorite)
+                if (favorites != null && favorites.contains(favRelative)) {
+                    val mFavorites = favorites.toMutableList()
```
Comment (kirillt) [inline arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Do we really need to convert to mutable structure here?

**Comment by mdrlzy** [comment]
Something is wrong with the removal. I got stack overflow when deleting a folder with 3 files.

<details>
  <summary>Log</summary>

```
java.lang.StackOverflowError: stack size 1039KB
        at sun.nio.fs.LinuxFileSystemProvider.getFileAttributeView(LinuxFileSystemProvider.java:74)
        at sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:144)
        at sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:115)
        at java.nio.file.Files.readAttributes(Files.java:1737)
        at java.nio.file.Files.isDirectory(Files.java:2192)
        at space.taran.arkfilepicker.folders.FoldersRepo.deleteFilesRecursively$lambda-1(FoldersRepo.kt:161)
        at space.taran.arkfilepicker.folders.FoldersRepo.$r8$lambda$023ki65JynAYimK3ciJu7Vv5xQg(Unknown Source:0)
        at space.taran.arkfilepicker.folders.FoldersRepo$$ExternalSyntheticLambda0.accept(Unknown Source:6)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:604)
        at space.taran.arkfilepicker.folders.FoldersRepo.deleteFilesRecursively(FoldersRepo.kt:160)
        at space.taran.arkfilepicker.folders.FoldersRepo.deleteFilesRecursively$lambda-1(FoldersRepo.kt:162)
        at space.taran.arkfilepicker.folders.FoldersRepo.$r8$lambda$023ki65JynAYimK3ciJu7Vv5xQg(Unknown Source:0)
        at space.taran.arkfilepicker.folders.FoldersRepo$$ExternalSyntheticLambda0.accept(Unknown Source:6)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:604)
        ...
```

</details>

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt [hunk 3]
```diff
@@ -119,11 +113,57 @@ class FoldersRepo(private val appCtx: Context) {
...
+    private fun deleteFilesRecursively(path: Path) {
+        val files = Files.list(path)
+        if (files != null) {
+            files.forEach {
+                if (it.isDirectory())
+                    deleteFilesRecursively(path)
+                else Files.delete(it)
+            }
+        }
+        Files.delete(path)
```
Comment (mdrlzy) [inline arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Maybe `path.toFile().deleteRecursively()`?

**Comment by mdrlzy** [comment]
There are too many icons in item right now
We need to replace the add/forget buttons with three dots icon and a popup.
You can use this [library](https://github.com/skydoves/Balloon) or [popup](https://github.com/ARK-Builders/ARK-Navigator/blob/main/app/src/main/java/space/taran/arknavigator/ui/view/DefaultPopup.kt) from navigator.

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt (hunk 1 above)
Comment (shubertm) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Alright, cool

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt (hunk 2 above)
Comment (shubertm) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Am not sure, but I guess immutable deletion can work too. Like your comment above :)

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt (hunk 3 above)
Comment (shubertm) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Thanks for the suggestion, I think it will solve the stackoverflow

**Comment by shubertm** [comment]
> There are too many icons in item right now We need to replace the add/forget buttons with three dots icon and a popup. You can use this [library](https://github.com/skydoves/Balloon) or [popup](https://github.com/ARK-Builders/ARK-Navigator/blob/main/app/src/main/java/space/taran/arknavigator/ui/view/DefaultPopup.kt) from navigator.

Alright

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt [hunk 4]
```diff
@@ -130,19 +128,31 @@ class FoldersRepo(private val appCtx: Context) {
         withContext(Dispatchers.IO) {
             if (folders.containsKey(root)) {
                 forgetRoot(root)
-                deleteFilesRecursively(root)
+                Log.d(
+                    "folders repo",
```
Comment (kirillt) [inline arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt:132]: For module logged, let's use constant defined in some common place.

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt (hunk 1 above)
Comment (shubertm) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Immutable deletion using ```minus``` is not working, so I reverted to mutable removal until we find a better way to perform immutable removal

Comment (shubertm) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt]: Immutable deletion using ```minus``` is not working, so I reverted to mutable removal until we find a better way to perform immutable removal

File: arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt (hunk 4 above)
Comment (mdrlzy) [reply arkfilepicker/src/main/java/space/taran/arkfilepicker/folders/FoldersRepo.kt:132]: Fixed in another PR

**Review by kirillt — APPROVED** [review]

**Pull request merged**